### DISCUSSION TITLE
Pass DotnetObjectReference to JS using [JSImport] API?

### DISCUSSION THREAD

**Discussion by Matheos96** [dotnet/aspnetcore#53866] [Q&A] [answered]
Is this possible? I doubt it myself a bit as my IDE is giving me an error:
![image](https://github.com/dotnet/aspnetcore/assets/32064310/5f6cf0e7-4083-4dcd-bdcb-e5cee8cfbad6)

Though, according to the docs 

> You can pass both JS and managed object references, and they are marshaled as proxy objects, keeping the object alive across the boundary until the proxy is garbage collected.

Though it is true that `DotnetObjectReference` is not listed in the table https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-8.0

Any way in making this work or will I need to keep using IJsInProcessRuntime service? 

BTW, I run blazor wasm standalone

**Comment by FLAMESpl** [comment]
Your issue has nothing to do with `DotNetObjectReference` type, or arguments of the method at all. If you had carefully read the error message you would know that partial method is missing its implementation. For every partial method in your project there must be a one that has an actual implementation. Dotnet SDK offers a source generator that creates implementations of JSImported partial methods, so in your case source generator was not triggered at all, which if I recall correctly, was a common issue in older versions of visual studio 2022 (or perhaps it is linked to SDK version?)

**Reply by Matheos96** [reply]
Hi. Well you are right in saying that the error message is clear, but my suspicion remains the same. 

The error is (if my guess is correct) indirectly due to the fact that the JSImport source generator is not able to generate an implementation in this case as it does not handle `DotNetObjectReference` type objects. The source generator works fine for other JSImport methods of obviously supported parameter types, which is why I think it simply won't work for this type. 

If anyone can get an example working with JSImport and, for example,  `DotNetObjectReference` as a parameter or return type, I will happily dismiss my theory.

In regards to SDK and VS versions, I have the latest VS 2022 version and latest .NET 8 SDK and, as I said, the source generator works fine for explicitly supported object types.

**Reply by javiercn** [reply, marked as answer]
You can't use `DotNetObjectReference` with the JSImport API. That's the native webassembly specific interop mechanism, not the Blazor one. The moment you opt-in to use `[JSImport]` you lock yourself into webassembly only and that interop model. 

Our recomendation is that unless you need to squeeze out the last drop of performance you stick with the Blazor interop model and only opt-in into the webassembly interop model as a progressive enhancement in the cases where you need it.

**Reply by Matheos96** [reply]
Thank you javier! This is exactly the answer I was suspecting, and waiting for. It is unfortunate but I understand there is no point in doing any further research into the topic.

We are currently working a hybrid of the two modes. We do want to squeeze out the last drop of performance, so we are doing everything in our power to make it happen. I may need to do some major refactoring if I want everything we have to work with JSImport (need to stop passing object references and, I dont know, use ids? something

Thanks :)
